File: src/api/watchlists.ts
import { callCommand } from "./tauri";
import type { Watchlist, WatchlistMembership } from "./types";

export type CreateWatchlistInput = {
  name: string;
  description: string | null;
};

export type RenameWatchlistInput = {
  id: string;
  name: string;
  description: string | null;
};

export type WatchlistMembershipInput = {
  watchlistId: string;
  companyId: string;
};

export function listWatchlists() {
  return callCommand<Watchlist[]>("list_watchlists");
}

export function listWatchlistMemberships() {
  return callCommand<WatchlistMembership[]>("list_watchlist_memberships");
}

export function createWatchlist(input: CreateWatchlistInput) {
  return callCommand<Watchlist>("create_watchlist", { input });
}

export function renameWatchlist(input: RenameWatchlistInput) {
  return callCommand<Watchlist>("rename_watchlist", { input });
}

export function deleteWatchlist(watchlistId: string) {
  return callCommand<void>("delete_watchlist", { watchlistId });
}

export function addCompanyToWatchlist(input: WatchlistMembershipInput) {
  return callCommand<void>("add_company_to_watchlist", { input });
}

export function removeCompanyFromWatchlist(input: WatchlistMembershipInput) {
  return callCommand<void>("remove_company_from_watchlist", { input });
}

File: src-tauri/src/commands/watchlists.rs
use crate::{app_state, storage};

#[tauri::command]
pub fn list_watchlists(
    state: tauri::State<'_, app_state::AppState>,
) -> Result<Vec<storage::Watchlist>, String> {
    state
        .watchlists()
        .list_watchlists()
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn list_watchlist_memberships(
    state: tauri::State<'_, app_state::AppState>,
) -> Result<Vec<storage::WatchlistMembership>, String> {
    state
        .watchlists()
        .list_watchlist_memberships()
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn create_watchlist(
    input: storage::NewWatchlist,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<storage::Watchlist, String> {
    state
        .watchlists()
        .create_watchlist(input)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn rename_watchlist(
    input: storage::WatchlistUpdate,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<storage::Watchlist, String> {
    state
        .watchlists()
        .rename_watchlist(input)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn delete_watchlist(
    watchlist_id: String,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<(), String> {
    state
        .watchlists()
        .delete_watchlist(&watchlist_id)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn add_company_to_watchlist(
    input: storage::WatchlistCompanyInput,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<(), String> {
    state
        .watchlists()
        .add_company_to_watchlist(input)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn remove_company_from_watchlist(
    input: storage::WatchlistCompanyInput,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<(), String> {
    state
        .watchlists()
        .remove_company_from_watchlist(input)
        .map_err(|error| error.to_string())
}
